Improve ipsec compile-time testing in CI

joestringer opened the pull request:
Improve the defines used in the BPF Makefile so that we typcially compile all of the encryption-related logic in the BPF programs when testing in the CI, so we can catch errors earlier.

joestringer committed:
bpf: Compile ipsec in tree

Previously, the default compile of bpf_network.o in the tree wouldn't
compile with IPv4, IPv6, or IPSEC enabled, which limited its ability to
catch compile failures at development time. Fix it up by adding these to
the BPF_SIMPLE_OPTIONS.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
bpf: Enable IP_POOLS in testing

Whenever we test ENABLE_IPSEC, also enable IP_POOLS which enables extra
code in most circumstances, except in one or two situations where it is
equivalent to the alternative (basically just determines which
skb->cb0[] offset to use for storing encryption bits).

Signed-off-by: Joe Stringer <joe@cilium.io>

pchaigno commented on bpf/Makefile:
```diff
@@ -87,7 +87,7 @@ MAX_BASE_OPTIONS = -DSKIP_DEBUG=1 -DENABLE_IPV4=1 -DENABLE_IPV6=1 \
 ifeq ("$(KERNEL)","49")
 # IPSec is incompatible with BPF NodePort so we only enable on 4.9.
-MAX_BASE_OPTIONS += -DENABLE_IPSEC=1
+MAX_BASE_OPTIONS += -DENABLE_IPSEC=1 -DIP_POOLS=1
```
Can `IP_POOLS` only be defined on 4.9?

jrfastab replied:
IP_POOLS is (or at least shortly will be) used on 4.14 kernels in EKS and 5.4 kernels in EKS. Also users can enable it almost anywhere by adding a subnet. There shouldn't be any kernel requirements around using it.

jrfastab replied:
There was a "bug" floating around where IP_POOLS without fib lookup helper would cause errors.

pchaigno replied:
Then it should be defined on line 87.

joestringer replied:
My approach here wasn't too complicated, I just put `IP_POOLS` wherever IPSec is enabled. Given that it's directly tied into that feature, that makes more sense to me rather than splitting it up to somewhere else. If we are only defining ENABLE_IPSEC on 4.9 due to the nodeport incompatibility then it'll make no difference to define IP_POOLS above.

Unless you're saying that the BPF nodeport incompatibility is no longer the case and we can drop the 4.9 kernel check entirely?

joestringer replied:
(sidenote, the naming of `IP_POOLS` sounds generic but it's really an IPSEC-specific config option so we could probably refactor that to make it clearer but I'm trying not to bit off too many changes at once ;) )

jrfastab replied:
also wait until we get backports sorted so we don't have to backport around naming.

pchaigno replied:
Got it, thanks for the explanation!

aditighag reviewed (commented):
Looks fine to me, but the change is more related to ipsec than loader. I'm inclined to remove my review assignment as Paul's review should cover both the areas.

joestringer committed:
bpf: Test build with -DHAVE_FIB_LOOKUP

Add extra build options with this to catch build-time errors
with/without this option. This is normally controlled by kernel version
support, but we don't currently factor variants of such feature
detection into the build testing.

Signed-off-by: Joe Stringer <joe@cilium.io>
